example_workspace/src/hello_world_example_app/app/src/main/java/com/ros/example/hello_ros/MainActivity.java

```diff
@@ -0,0 +1,159 @@
+    private Button runButton;
+    private TextView statusText;
+    private Status status;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        Log.i(TAG, "OnCreate()");
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.logging);
+
+        masterIP = (EditText) findViewById(R.id.master_ip);
+        myIP = (EditText) findViewById(R.id.my_ip);
+        runButton = (Button) findViewById(R.id.run_button);
+        statusText = (TextView) findViewById(R.id.status);
+        status = Status.WAITING;
+
+        Log.i(TAG, "new RosThread");
+        mainThread = new RosThread();
+        Log.i(TAG, "end RosThread constructor");
+
+        runButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                runButtonCallback();
+            }
+        });
+
+        String ip = getLocalIpAddress();
+        if (ip != null) {
+            myIP.setText(ip);
+        }
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        if (status == Status.RUNNING) {
+            new Thread(mainThread).start();
+        }
+    }
+
+    @Override
+    protected void onPause() {
+        super.onPause();
+        mainThread.stop();
+    }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        mainThread.stop();
+    }
+
+    private void runButtonCallback() {
+        String ipRegexExpression =
+            "^(([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.){3}([01]?\\d\\d?|2[0-4]\\d|25[0-5])$";
+
+        switch (status) {
+            case WAITING:
+                String sMasterIP = masterIP.getText().toString();
+                if (! sMasterIP.matches(ipRegexExpression)) {
+                    statusText.setText(R.string.status_masterIP_error);
+                    break;
+                }
+                Log.i(TAG, "master ip is fine: " + sMasterIP);
+
+                String sMyIP = myIP.getText().toString();
+                if (! sMyIP.matches(ipRegexExpression)) {
+                    statusText.setText(R.string.status_myIP_error);
+                    break;
+                }
+                Log.i(TAG, "my ip is fine: " + sMyIP);
+
+                if (! mainThread.checkRosMaster(sMasterIP, sMyIP)) {
+                    statusText.setText(R.string.status_check_master_error);
+                    break;
+                }
+                Log.i(TAG, "Master is ready");
+
+                statusText.setText(R.string.status_running);
+                runButton.setText(R.string.button_stop);
+
+                new Thread(mainThread).start();
+                status = Status.RUNNING;
+                break;
+
+            case RUNNING:
+                statusText.setText(R.string.status_waiting);
+                runButton.setText(R.string.button_run);
+
+                mainThread.stop();
+                status = Status.WAITING;
+                break;
+        }
```

Comment:
As far as I know `roscpp` does not allow to initialize the node twice, so `ros::init()` cannot be called twice within the same process, even if you call `ros::shutdown()` in between. Given that this is true, does it make sense to allow to toggle the state between `WAITING` and `RUNNING`?

Reply:
See https://github.com/Intermodalics/ros_android/pull/101#discussion_r251577050; this seems to be working.

Reply:
It cannot if `g_initialized` is not destroyed and reinitialized to its default value: It is only ever set to `true` in [init.cpp:469](https://github.com/ros/ros_comm/blob/17d8c67b6d79cd44d8c3f3675b77787ee93a2ddd/clients/roscpp/src/libros/init.cpp#L469) and never reset. It might seem to work, but the new settings won't take effect. This behavior might be trivial to fix in roscpp, but without a patch you cannot change the master URI or ROS_IP without terminating the app or unloading all shared libraries.

Reply:
I see. So this would work only to start / stop the program using the same ROS master; in any case it still sounds somewhat useful to me. I'll add a note somewhere clarifying that the app should be restarted if the ROS master changes after the first connection.

Reply:
I tested the app today and as expected changes of the ROS master IP have no effect until the app is terminated by swiping it up or down in the app view.